File: medium/86-partition-list/brute.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func partition(head *ListNode, x int) *ListNode {
	var lc, rc, lhead, rhead *ListNode

	cursor := head
	for cursor != nil {
		theNext := cursor.Next
		// cursor.Next = nil

		if cursor.Val < x {
			if lc == nil {
				lc = cursor
				lhead = cursor
			} else {
				lc.Next = cursor
				lc = lc.Next
			}
		} else {
			if rc == nil {
				rhead = cursor
				rc = cursor
			} else {
				rc.Next = cursor
				rc = rc.Next
			}
		}

		cursor = theNext
	}

	if lc == nil {
		return rhead
	}

	lc.Next = rhead
	return lhead
}

func main() {
	head := &ListNode{Val: 2}
	one := &ListNode{Val: 1}
	head.Next = one
	h := partition(head, 2)
	for h != nil {
		fmt.Println(h.Val)
	}
}

File: medium/86-partition-list/brute.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        if (head == nullptr) {
            return head;
        }
        ListNode* lc = nullptr;
        ListNode* rc = nullptr;
        ListNode* lhead = nullptr;
        ListNode* rhead = nullptr;

        ListNode* cursor = head;

        while (cursor != nullptr) {
            ListNode* theNext = cursor->next;
            cursor->next = nullptr;

            if (cursor->val < x) {
                if (lc == nullptr) {
                    lc = cursor;
                    lhead = cursor;
                } else {
                    lc->next = cursor;
                    lc = cursor;
                }
            } else {
                if (rc == nullptr) {
                    rhead = cursor;
                    rc = cursor;
                } else {
                    rc->next = cursor;
                    rc = cursor;
                }
            }

            cursor = theNext;
        }
        if (lc ==  nullptr) {
            return rhead;
        }

        lc->next = rhead;
        return lhead;
    }
};